_posts/2015-09-12-developing-gamepad-virtual-keyboard.markdown

```diff
@@ -0,0 +1,150 @@
+
+CSS classes or selective JS rendering can handle the Viewport visibility. And Flux can help us out here by tracking whether the viewport should be visible. As I mentioned earlier, a Flux store is also a good place to keep the "current" text value. Updating it when a new value is inherited from the Real Input.
+
+With Viewport visibility taken care of, resizing is the next challenge. I ended up using a combination of JS and CSS to keep the Viewport fullscreen and centered. I wrote CSS that fit the Viewport on screen perfectly for a particular screen size, then I manually scaled its proportions using JS and the `scale()` CSS property. I highly recommend this approach because there is no degradation in quality when using CSS transforms like scale(). And because DaisywheelJS doesn't use any images and only uses the DOM to make it's shapes, everything scales crisply. Below is some of the DaisywheelJS’s Modal (its Viewport) React component code.
+
+```js
+
+// jsx/Modal.js
+
+setupSize: _.throttle(function() {
+    var innerHeight = window.innerHeight;
+    var innerWidth = window.innerWidth;
+    var maxSize = (innerWidth < innerHeight) ? innerWidth : innerHeight;
+    var padding = 25;
+    var scalePrecision = 100;
+    var modalHeight = 1000;
+    var scaleAmount = Math.floor(scalePrecision * (maxSize / (modalHeight + padding*2))) /    scalePrecision;
+
+    this.setState({
+        scale: scaleAmount
+    });
+}, 100),
+
+componentDidMount: function() {
+    this.setupSize();
+},
+
+componentWillMount: function() {
+    window.addEventListener('resize', _.bind(this.setupSize, this));
+},
+
+componentWillUnmount: function() {
+    window.removeEventListener('resize', _.bind(this.setupSize, this));
+},
+
+```
+
+### The controls
+
+This is another part of the interface that depends on implementation. In this article we've discussed implementing Backspace and using the DPad, so the UI should at least reflect those. A Space button is also something to consider, and it's easy to implement.
+
+The larger challenge comes from including multiple character sets. The most obvious being "capital characters", but this might also includes sets for other languages, and in DaisywheelJS's case, [even Emoji Fonts](http://daisywheeljs.org/#options-and-extensions). Toggling or cycling through character sets should be controlled through some button as well, so we'll add that to the list of controls. Tracking the currently selected symbol is another one of those pieces of state that is so easily managed through Flux. Since the keyboard view needs to know what characters to show and the Fake Input needs to know what character was just selected, a store is the perfect place to manage the character sets.
```

Comment:
The larger challenge comes from including multiple character sets, the most obvious being "capital characters". This might also includes sets for other languages and, in DaisywheelJS's case, [even Emoji Fonts](http://daisywheeljs.org/#options-and-extensions).